MockRealtimeVideoSource creates a ref cycle

youennf opened the pull request:
#### ef713209045c06d14b7a92c69ac1d49d6a5e049f
<pre>
MockRealtimeVideoSource creates a ref cycle
<a href="https://bugs.webkit.org/show_bug.cgi?id=284817">https://bugs.webkit.org/show_bug.cgi?id=284817</a>
<a href="https://rdar.apple.com/problem/141616762">rdar://problem/141616762</a>

Reviewed by Chris Dumez and Phil Normand.

Before the patch MockRealtimeVideoSource would never be destroyed as one of its member, m_emitFrameTimer, was taking a ref to it without ever releasing it.
Instead, we are using a different constructor for m_emitFrameTimer which will only ref MockRealtimeVideoSource when actually using it.

* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::m_emitFrameTimer):

Canonical link: <a href="https://commits.webkit.org/288006@main">https://commits.webkit.org/288006@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1557fb0fd866e25dde2040045dc478ed59fe17d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81675 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1200 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35624 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86215 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32665 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1231 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9019 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63715 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21441 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84745 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/881 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74325 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44007 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/780 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28502 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31120 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29101 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87654 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8912 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72055 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9095 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70150 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71290 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17746 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15368 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14283 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8864 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8704 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12225 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10513 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
```diff
@@ -118,6 +118,12 @@
+void MockRealtimeVideoSourceMac::endProducingData()
+{
+    MockRealtimeVideoSource::stopProducingData();
+    stopRunLoop();
```
Could you add this in the GStreamer mock source too? I suspect we have the same issue there.

cdumez reviewed (approved):
LGTM

philn reviewed (approved):
Even simpler, thanks!

philn commented on Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
```diff
@@ -83,6 +83,7 @@ class MockRealtimeVideoSource : public RealtimeVideoCaptureSource, private Orien
 
     void startProducingData() override;
     void stopProducingData() override;
+    void endProducingData() final;
```
This is source of asserts here, I haven't found why...

youennf replied:
I'll revert that part for now.

youennf committed:
MockRealtimeVideoSource creates a ref cycle
https://bugs.webkit.org/show_bug.cgi?id=284817
rdar://problem/141616762

Reviewed by Chris Dumez and Phil Normand.

Before the patch MockRealtimeVideoSource would never be destroyed as one of its member, m_emitFrameTimer, was taking a ref to it without ever releasing it.
Instead, we are using a different constructor for m_emitFrameTimer which will only ref MockRealtimeVideoSource when actually using it.

* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::m_emitFrameTimer):

Canonical link: https://commits.webkit.org/288006@main
